gemonlinedb/src/common/XMLConfigurationProvider.cc

```diff
@@ -0,0 +1,203 @@
+            template<>
+            std::string toString<PartReferenceSN>(
+                const PartReferenceSN &reference)
+            {
+                return reference.serialNumber;
+            }
+
+            /// @brief Loads and validates a document.
+            DOMDocumentPtr loadDOM(const std::string &filename,
+                                   const std::string &searchPath,
+                                   const std::string &schema)
+            {
+                using namespace detail::literals;
+
+                // Find the file (throws if not found)
+                auto path = detail::getFileInPath(filename, searchPath);
+
+                // Find the schema (throws if not found)
+                auto schemaPath = detail::getFileInPath(
+                    schema, "xml/schema:/opt/cmsgemos/share/gemonlinedb/xml/schema");
+
+                // Make the paths absolute
+                path = boost::filesystem::canonical(path).string();
+                schemaPath = boost::filesystem::canonical(schemaPath).string();
+
+                // Get a DOM implementation
+                auto impl = DOMImplementationRegistry::getDOMImplementation("LS"_xml);
+                if (impl == nullptr) {
+                    throw std::logic_error("Cannot find a DOM implementation");
+                }
+
+                auto errorHandler =
+                    std::make_shared<detail::XercesAlwaysThrowErrorHandler>();
+
+                XercesDOMParser parser;
+                parser.setErrorHandler(errorHandler.get());
+                parser.setExternalNoNamespaceSchemaLocation(
+                    detail::XercesString(schemaPath));
+                parser.setValidationScheme(XercesDOMParser::Val_Always);
+                parser.setDoNamespaces(true);
+                parser.setDoSchema(true);
+                parser.setValidationSchemaFullChecking(true);
+
+                // Parse and validate
+                parser.parse(detail::XercesString(path));
+
+                DOMDocumentPtr document;
+                document.reset(parser.adoptDocument());
+                return document;
+            }
+
+            /// @brief Common part to all loadXXX methods
+            template<class ConfigurationType>
+            void loadInternal(
+                const std::string &filename,
+                const DOMDocumentPtr &document,
+                std::map<std::string, std::shared_ptr<ConfigurationType>> &map)
+            {
+                XMLSerializationData<ConfigurationType> xmlData;
+                xmlData.readDOM(document);
+
+                auto dataSets = xmlData.getDataSets();
+                for (const auto &set : dataSets) {
+                    auto part = set.getPart();
+                    if (map.count(toString(part)) > 0) {
+                        // Corresponds to more than one configuration for the same
+                        // part, which is not supported.
+                        throw std::runtime_error(
```

Comment:
I would rather define custom exceptions for such things (anyway they ave to be defined for the cases outlined above)

Reply:
Yes, all `xdaq` related code should throw (a derivative of) `xcept:exception`, you can use the on-the-fly exception creation macros in the, e.g., `include/gem/onlindb/exception/Exception.h` header, or define specific exceptions there.

Reply:
Already fixed in 8562f2918b694c9ce6fb1c809b6c043e43af0ee7